### DISCUSSION TITLE
Package FINN IP's from Stitch Project

### DISCUSSION THREAD

**Discussion by smtalds** [Xilinx/finn#885] [Q&A] [answered]
Hi , i want to deployment finn's ip (like , streaming model , idma and odma ) to another block design.I opened these ip from stitch project and packaged. I recalled and added ip to my block design. But when i want to validate design , gives this error : 

![image](https://github.com/Xilinx/finn/assets/59387657/c44177f0-f52a-4c0b-b0dd-d97087810c18)

What am i missing ?

**Comment by fpjentzsch** [comment]
Hi,
did you manage to find the problem? My only idea is that "invalid target" refers to an incorrectly specified FPGA part or a part that is incompatible with the one set in the FINN build flow.

**Reply by smtalds** [reply, marked as answer]
Yes I found the problem. I was getting IPs from a non-ip shared folder. I was trying to use packed IP. When I use ip shared verilog files the problem was solved. (It was so long time ago, i maynot remember)

**Reply by JPPalacios** [reply]
Hi,

Would you mind walking me through how you integrated your IP into a larger design? Trying to do the same but adding it to the PYNQ base overlay. I have the FINN flow working on hardware so I am simply trying to make PS stuff in the PL like the input end.

Regards,

**Reply by smtalds** [reply]
I don't remember exactly, but you might be able to use the design in the ip-shared folder. Take all the Verilog files and add your design, but don't use the closed IP. Alternatively, you can change the training bitwidth. When I was using the Z1, I would decrease the bitwidth to get more FPS. If the FPGA is the same, change your IP design. If the IP is the same, then change your FPGA.

I couldn't find any solution in November. Maybe FINN has different updates now. Can you check?

**Reply by fpjentzsch** [reply]
Maybe this tutorial can help? https://github.com/Xilinx/finn/tree/dev/tutorials/fpga_flow

**Reply by JPPalacios** [reply]
@smtalds @fpjentzsch Thanks, I was able to generate the Stitched IP. If I do integrate the FINN IP into a larger design, would the FINN Example Overlay class still work? The input end would directly receive MIPI data into the streaming FIFOs instead of it being pre-processed in the PS. The outputs would ideally go to the PS for post-processing before sending frames off to the DP or a matplotlib window. Target board is the Kria KV260.
 
Note: I've tried the Kria-PYNQ notebooks but there are no working Base Overlay MIPI examples unfortunately, unresolved I2C issues. The smartcam app works just fine on my kria board though so I know this type of work is possible, just prefer if FINN was handling detection over the DPU :) 

Thanks for the help once again!

**Reply by fpjentzsch** [reply]
You could re-use parts of the FINN Example Overlay (or driver_base.py) if you instantiate both or one of the FINN-generated DMAs (the output DMA in your case) in your block design. For example, you can stream the data into the accelerator via some other source (your MIPI) but use the output DMA and corresponding FINN driver functions to continuously read the accelerator output.

The DMAs FINN instantiates for use with the integrated FINN Example Overlay driver are not part of the stitched IP. But you should be able to extract the IP cores from your (temporary) FINN BUILD DIR when you do a full "ZynqBuild". There will probably be 3 "StreamingDataflowPartition". Numer 0 is the input DMA, 1 is the FINN accelerator itself (stitched IP), 2 is the output DMA (not 100% sure about the ordering).

Alternatively, you could use a generic DMA core from Vivado and control it with the dma driver included in PYNQ.

**Reply by smtalds** [reply]
@JPPalacios, Yes its works. But don't forget that Input size and your data should be same. Other way, you may get false output on DMA. You can find DMA base address on driver python files. You should check files.

**Reply by JPPalacios** [reply]
Hi @fpjentzsch @smtalds ,

I was able to integrate the FINN into a larger design with good results. I now want to remove the `idma0` IP and stream data directly into `StreamingDataflowPartition_1`. I have seen that I may need to modify the FINN transformations to include a TLAST marker. Does this still apply to the latest version of FINN? The `odma0` would remain the same in this design for PS post-processing.

Thank you for your time

**Reply by smtalds** [reply]
Hi @JPPalacios,
I'm not sure how you can do that because these signals are generated automatically by the FINN scripts. The DMA helps with data management and pipeline. Why do you need this? If you need to increase FPS, I guess DMA does not affect your FPS, but I am not sure.

**Reply by JPPalacios** [reply]
@smtalds I do some preprocessing to my data such as letterboxing. I would like to capture a frame, broadcast this stream to the VDMA _and_ to the FINN `StreamingDataflowPartition_1`. I hear getting AXI Streams to work with an AXI broadcast is a little tricky so I may be out of luck.

**Reply by smtalds** [reply]
Hi @JPPalacios, If i understand you correctly, you change dma by `VDMA`Did you see block design of `idma&dataflow&odma`? Change and connect signals. Maybe you have to change base address vdma with same dma base address. I was used yolo with FINN. I did not change DMAs. I get frame and write to DMA if i remember correct. You have to use `TData`, `tvalid` and `tready`. If your question is different from this, we can discuss.

**Reply by JPPalacios** [reply]
Hey @smtalds,

I have design `I` implemented and working good. In my attempt to minimize the number of PS/PL data transfers, I was curious if I could just use AXI Streaming for pre-processing a 720p into the correct shape for my `dataflow` block.

![image](https://github.com/user-attachments/assets/695a45a6-b508-4831-b60c-41f549899f51)

I am interested in something like `II` or `III`.

**Reply by smtalds** [reply]
Hi @JPPalacios ,

You already have a `VDMA`, which is a type of` DMA.` So, you have **`AXI signals`**. Your `HLS Letterbox` is connected to the `AXI DMA`, meaning it also has `AXI signals`.

Given the presence of AXI signals, you can connect the output of the `VDMA` to the `HLS Letterbox`. Then, connect the output of the `Letterbox` to an `AXI FIFO`, and finally connect the output of the FIFO to` the dataflow.` 

However, this is just my opinion. You should verify the data. Perhaps you can use regular data like counter data to test. If you get the correct data at the output of the AXI FIFO, then our design has a good pipeline. End of the day, we romoved `AXI-DMA` and `IDMA`. We added an `AXI FIFO`. Our PL pipeline is shorter now.

If  your design is not special, i want to see. I guess your using tiny-yolo. How much fps you get? When i was working on, i got like 10-15fps a frame with PYNQ Z1. I did not use new FINN framework. It was old.

**Reply by JPPalacios** [reply]
Hey @smtalds,

The FINN design I [referenced](https://github.com/sefaburakokcu/finn-quantized-yolo) benchmarked at 15-20 fps which sources video from a webcam/PS letterboxing. Using the PYNQ framework, we can speed this up quite a bit so I am at around 25-30 fps. I haven't changed the actual layer design from the ref design except use FINN's RTL version in a build session. I should mention that I am using a Kria KV260 but perhaps in a future version, I'd like to target a smaller design.

Thanks for the suggestions, I will see if I can implement the vdma write channel portion!

**Reply by smtalds** [reply]
Hi, @JPPalacios. I was used this repo too. If you can do that i want to analyze. You can change anchor numbers. I detect object with more fps but i lost a little accuracy and detection performance. #https://github.com/ultralytics/yolov5/issues/11975

**Reply by JPPalacios** [reply]
Hi @smtalds,

I am trying to re-train the network for 1w1a but get an assertion error triggered by `ConvertBipolarMatMulToXnorPopcount` in `bipolar_to_xnor.py:94`:

```python
# this code is in bipolar_to_xnor.py
# ensure old scale/bias were correct for BIPOLAR
scale_ok = mt_inst.get_nodeattr("out_scale") == 2.0
bias_ok = mt_inst.get_nodeattr("out_bias") == -1.0
assert (scale_ok and bias_ok)
```

```bash
 line 94, in apply
    assert (
AssertionError: Unexpected scale/bias attributes for BIPOLAR MultiThreshold node. out_scale: 1.0, out_bias: 0.0
``` 

The last intermediate model before the assertion error looks like this: 

![image](https://github.com/user-attachments/assets/ccfd2af1-062d-4060-a0e9-f277867bf424)

Did you face any issues when reducing the bit widths to 1w1a, 2w2a, 1w2a, or 2w1a models? I can't seem to find anyone else working on this particular model so I thought I'd ask you.

Thanks

**Reply by smtalds** [reply]
Hi, @JPPalacios, which version of FINN and Brevitas are you using? I trained with 1w1a,2w2a and more configuration. Maybe you get this error because of new version of FINN. Your inputs is not integer. Maybe you can use [RoundAndClipThresholds](https://github.com/Xilinx/finn/blob/main/src/finn/transformation/streamline/round_thresholds.py#L33)  class?

And maybe @fpjentzsch can help us :)

**Reply by JPPalacios** [reply]
Hey @smtalds,

I am using the latest versions. I don't use any other custom steps besides merging the [pre-processing node](https://medium.com/@bestamigunay1/end-to-end-deployment-of-lpyolo-low-precision-yolo-for-face-detection-on-fpga-13c3284ed14b) at the top. When I compare an intermediate model from the same `step_tidy_up` step in a working finn-build, I notice there are [Mul](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Mul) nodes after the `MultiThreshold` nodes. Found some more information in this [thread.](https://github.com/Xilinx/finn/issues/1031#issuecomment-2044629046)

Would you mind sharing how you implemented the build step? Perhaps I may have to include custom steps to handle 1w1a models. I could not find a tutorial on this. Perhaps @fpjentzsch  may point us in the right direction. I truly appreciate your times and guidance.

**Reply by smtalds** [reply]
Hi @JPPalacios, i was used FINN 0.7 version because there are shared [this](https://github.com/sefaburakokcu/finn-quantized-yolo#:~:text=Requirements-,Finn%20%3D%3D%200.7,-Pytorch%20%3E%3D%201.8.1). I did not change anything like custom steps. You may use this version. Let wait to answer from @fpjentzsch.

And i want to share something about you. [Yolo-v10](https://docs.ultralytics.com/models/yolov10/#architecture:~:text=Key%20Features-,NMS%2DFree%20Training,-%3A%20Utilizes%20consistent%20dual) is released without NMS. If you are not busy maybe you can try this version. You may get better fps from PS because of without NMS.

**Reply by JPPalacios** [reply]
Hey @smtalds, I have been staying away from nonlinear topologies as I am not sure they're supported yet (I.e., skip connections). Perhaps this has changed or some tutorial that explains the types of transformations required for a build. Let me know if you take a crack at it.

For NMS, I was curious to find a way to implement this using HLS. Seems like it's something worth looking into. Otherwise, maybe creating custom QONNX nodes + HW Custom Ops.

**Reply by JPPalacios** [reply]
HI @smtalds ,

Did you ever train for more than a single class? If so, did you ever face issues with boxes during pynq-inference? When I run detect.py, I can see my .pt weights do okay but after exporting and finn-compilation, all my predictions are "zeroes" meaning the bounding boxes do not move and remain at the top of the image. Just thought I'd ask, no leads yet. Thanks!

**Reply by smtalds** [reply]
Hi @JPPalacios, 
Actually i dont remember. I trained for cars and person. Cars always detected but for person, ıhh i dont remember. Yes, detect.py always working but after the finn-compilation we get always problems. Did you increase weights-bias bit size? You can debug with jupyter  notebook. Do you have enough memory space? I dont know what cause is the problem.

**Reply by JPPalacios** [reply]
Hey @smtalds,

I did not increase weights (2w4a). Perhaps there's not enough bits to distinguish between more a single class. In a single class, perhaps it's easier to predict (face or not). This is anecdotal but I trained a model on 2 classes and saw predictions scatter/randomly placed, but not zeroes. For 3 classes, all predictions (y-coords) are set to zero. I am still leaning towards a hardware issue, though. I will attempt to load a larger model (4w4a) to see if this is something to consider for multi-class cases. Were you able to simulate your ONNX file? If so, how did you go about doing so? The examples sims are a little hard to follow.

**Reply by smtalds** [reply]
Hey @JPPalacios 

Multiple class should be work, because of  i worked with multiple class for sound and binary image for classification. I checked with netron. I did not simulate.

**Reply by JPPalacios** [reply]
> Hey @JPPalacios 
> 
> 
> 
> Multiple class should be work, because of  i worked with multiple class for sound and binary image for classification. I checked with netron. I did not simulate.

Good point, all vision examples are multi-class classifiers. I trained a single-class (cars) 4w4a LP-YOLO detector. The model outputs on the Kria are random (but not zero!) which implies my pipeline is at fault. Not sure where else to look, though.

**Reply by smtalds** [reply]
> Good point, all vision examples are multi-class classifiers. I trained a single-class (cars) 4w4a LP-YOLO detector. The model outputs on the Kria are random (but not zero!) which implies my pipeline is at fault. Not sure where else to look, though.

Are u using jupyter? I printed outputs on jupyter output console and debug. I trained 2w2a, 2w3a,3w2a and more combinations and get outputs. You can generate outputs for PYNQ. If you get true outputs with PYNQ you can use ipshared folder and move rtl files. Then you can generate bitstream file with Vivado. I generated for PYNQ Z1and did not get any wrong answer for xWxA combination. I dont be sure, maybe finn generation for Kria is false. What is the main difference, i guess nothing. Only we are used different board and get for this board finn generation.

@auphelia Hi, Have you encountered any problems while taking Finn synthesis for Kria? Thanks for answering.

**Reply by JPPalacios** [reply]
Hi @smtalds,

After going back through each of the intermediate models, I noticed I was missing a node on the custom one vs the LP-YOLO's model. I used the [original folding json file](https://medium.com/@bestamigunay1/end-to-end-deployment-of-lpyolo-low-precision-yolo-for-face-detection-on-fpga-13c3284ed14b) and that fixed the issue. I can now deploy single and multi (person + car) obj detectors. Thanks for your help and brainstorming! Thanks also to the FINN team for their hard work @fpjentzsch @auphelia. You guys rock!

**Reply by smtalds** [reply]
Hi @JPPalacios,

It's nice that you managed to execute it. Sometimes we miss certain things. When I read this comment, you mentioned that you made a 9 ms inference, right? Perfect. What are you doing for this

**Reply by JPPalacios** [reply]
Hi @smtalds,

For my config, using the RTL backends. I will have to check the timings again but somehow having a 4-bit model made things a little faster. I am going to revise my timings using `time.perf_time` for more accuracy.

**Reply by smtalds** [reply]
Nice works, you can change anchor size. I [changed ](https://github.com/ultralytics/yolov5/issues/11975) anchor like this. We can sacrifice accuracy but we can detect faster. You should try it.

What does it mean RTL backends? When i was using finn, it was a year ago. What is the changed?

**Reply by JPPalacios** [reply]
hey @smtalds,

This change makes it possible to use RTL IPs instead of the HLS versions. Each node now may have one or more "backends". You can now specify what each layer backend should be using the "specialize_layers.json" config file on top of the "folding_config.json" file. Thank you for reminding me of the anchor trick. I will see if this improves throughput.

**Comment by JPPalacios** [comment]
@fpjentzsch Hi,

I imported the Kria-PYNQ MIPI hierarchy block into the FINN Zynq build Vivado project (with i/o dmas + streamdataflow1) and synthesized the design. I had a few questions about generating hardware files, since I am running into implementation errors.

- After synthesis/bitstream generation, how does FINN generate the XDC/HWH files?
- What files would I need to modify to streamline this process for future iterations?

Many thanks for the support!

**Reply by fpjentzsch** [reply]
FINN takes the bitfile from the location specified here: https://github.com/Xilinx/finn/blob/e3087ad9fbabcc35f21164d415ababec4f462e9f/src/finn/transformation/fpgadataflow/make_zynq_proj.py#L267
and .hwh file from one of the locations specified here: https://github.com/Xilinx/finn/blob/e3087ad9fbabcc35f21164d415ababec4f462e9f/src/finn/transformation/fpgadataflow/make_zynq_proj.py#L276

That is all information needed for the "FINN Example Overlay" driver (along with the "io_shape_dict").

**Reply by JPPalacios** [reply]
Hi, I am exploring importing the FINN ZYNQ project (idma, streamdataflow1, odma) into Kria-PYNQ's Base Overlay. I am working out the details on merging the two first by GUI and then by write_bd_tcl command. 

This method will most likely use PYNQ custom IP/Hierarchies per their [tutorials](https://pynq.readthedocs.io/en/v3.0.0/overlay_design_methodology/overlay_tutorial.html#Overlay-Customisation) to bind to FINN IP's. I believe this is where I transfer the functionality of the FINNExampleOverlay into their corresponding blocks, correct? Perhaps this is outside FINN's scope but hoping you may be able to shed some light on this.

Regards,
